nuxeo-features/nuxeo-platform-comment/nuxeo-platform-comment/src/main/java/org/nuxeo/ecm/platform/comment/impl/CommentsMigrator.java

```diff
@@ -59,64 +74,166 @@
 
     protected static final int BATCH_SIZE = 50;
 
+    /**
+     * @since 11.1.
+     */
+    public static final String GET_COMMENTS_FOLDERS_QUERY = String.format(
+            "SELECT %s FROM Document WHERE %s = '%s' AND %s ='%s'", ECM_UUID, ECM_NAME, COMMENTS_DIRECTORY,
+            ECM_PRIMARYTYPE, HIDDEN_FOLDER_TYPE);
+
     @Override
     protected String probeSession(CoreSession session) {
-        CommentService commentComponent =  Framework.getService(CommentService.class);
+        CommentService commentComponent = Framework.getService(CommentService.class);
         CommentServiceConfig commentServiceConfig = commentComponent.getConfig();
         if (commentServiceConfig != null) {
             Graph graph = Framework.getService(RelationManager.class).getGraph(commentServiceConfig.graphName, session);
-            if (graph.getStatements().size() > 0) {
+            if (!graph.getStatements().isEmpty()) {
                 return MIGRATION_STATE_RELATION;
             }
         }
-        return MIGRATION_STATE_PROPERTY;
+        // If not in relation, check if there are still property comments under hidden Comments folder(s)
+        // Do a new query one to check the step and the second the make the migration (do all comments)
+        if (hasUnsecuredComments(session)) {
+            return MIGRATION_STATE_PROPERTY;
+        }
+        // Else, comments are already secured
+        return MIGRATION_STATE_SECURED;
     }
 
-    @Override
-    protected void migrateSession(CoreSession session) {
+    /**
+     * @since 11.1
+     */
+    protected List<DocumentModel> getUnsecuredComments(CoreSession session) {
+        List<String> parentIds = getCommentFolders(session);
+        if (parentIds.isEmpty()) {
+            return Collections.emptyList();
+        }
+        return session.query(buildAllCommentsQuery(parentIds, COMMENT_DOC_TYPE));
+    }
+
+    /**
+     * @since 11.1
+     */
+    protected boolean hasUnsecuredComments(CoreSession session) {
+        List<String> folderIds = getCommentFolders(session);
```

Comment:
This may be costly to probe the session. Here we just interested in the count whereas we fetch all comments containers. This is not so bad for `PropertyCommentManager` but it could be very costly for `CommentManagerImpl` which could have a container per comment at most (one container per month).

Reply:
actually the `org.nuxeo.ecm.platform.comment.impl.CommentsMigrator#getCommentFolders` returns the folders **Ids** (we use projection query) and if one of this folder contains children we stop iterating.

if we take the example of one comment per month we will have 12 folders per year.  what is for you the maximum threshold that be critical.